pkg/controller/directvolumemigration/resource.go

```diff
@@ -0,0 +1,352 @@
+	var podLimitRangeItem *corev1.LimitRangeItem
+	var containerLimitRangeItem *corev1.LimitRangeItem
+	for _, limitRange := range limitRangeList.Items {
+		for i := range limitRange.Spec.Limits {
+			limitItem := &limitRange.Spec.Limits[i]
+			switch limitItem.Type {
+			case corev1.LimitTypeContainer:
+				if containerLimitRangeItem == nil {
+					containerLimitRangeItem = limitItem
+				} else {
+					mergeResourceLimitRangeItems(containerLimitRangeItem, limitItem)
+				}
+			case corev1.LimitTypePod:
+				if podLimitRangeItem == nil {
+					podLimitRangeItem = limitItem
+				} else {
+					mergeResourceLimitRangeItems(podLimitRangeItem, limitItem)
+				}
+			}
+		}
+	}
+	if podLimitRangeItem != nil {
+		finalLimitRange.Spec.Limits = append(finalLimitRange.Spec.Limits, *podLimitRangeItem)
+	}
+	if containerLimitRangeItem != nil {
+		finalLimitRange.Spec.Limits = append(finalLimitRange.Spec.Limits, *containerLimitRangeItem)
+	}
+	return finalLimitRange, nil
+}
+
+// mergeResourceLimitRangeItems given target and source LimitRangeItems, merges source into target.
+// it compares the Max & Min resource limits of CPU and memory and keeps the most restrictive value
+// NOTE: does not consider MaxLimitRequestRatio for comparison
+func mergeResourceLimitRangeItems(target *corev1.LimitRangeItem, source *corev1.LimitRangeItem) {
+	if target != nil && source != nil {
+		if source.Max != nil {
+			// for max limits, the most restrictive value is the one which is the minimum
+			if target.Max.Cpu().Cmp(*source.Max.Cpu()) > 0 {
+				target.Max[corev1.ResourceCPU] = source.Max[corev1.ResourceCPU]
+			}
+			if target.Max.Memory().Cmp(*source.Max.Memory()) > 0 {
+				target.Max[corev1.ResourceMemory] = source.Max[corev1.ResourceMemory]
+			}
+		}
+		if source.Min != nil {
+			// for min limits, the most restrictive value is the one which is the maximum
+			if target.Min.Cpu().Cmp(*source.Min.Cpu()) < 0 {
+				target.Min[corev1.ResourceCPU] = source.Min[corev1.ResourceCPU]
+			}
+			if target.Min.Memory().Cmp(*source.Min.Memory()) < 0 {
+				target.Min[corev1.ResourceMemory] = source.Min[corev1.ResourceMemory]
+			}
+		}
+	}
+}
+
+// applyLimitRangeItemOnRequirements given a ResourceRequirement and a LimitRangeItem, applies limits on the given ResourceRequirement
+// the ResourceRequirement is modified in-place such that no resource value violates the limits in the given LimitRangeItem
+func applyLimitRangeItemOnRequirements(requirements *corev1.ResourceRequirements, limit corev1.LimitRangeItem) {
+	if requirements.Requests != nil {
+		if limit.Max != nil {
+			//  apply Max limits on Requests and Limits, none should exceed the Max
+			if requirements.Requests.Cpu().Cmp(*limit.Max.Cpu()) > 0 {
+				requirements.Requests[corev1.ResourceCPU] = *limit.Max.Cpu()
+			}
+			if requirements.Requests.Memory().Cmp(*limit.Max.Memory()) > 0 {
+				requirements.Requests[corev1.ResourceMemory] = *limit.Max.Memory()
+			}
+		}
+		if limit.Min != nil {
```

Comment:
If the namespace has non-intersecting limit ranges (i.e. one has a max of "5" and another has a min of "6"), the above won't detect the problem and just respond with requirements based on the minimums. Of course if those limit ranges exist, it will be impossible to start *any* pod in that namespace (even outside of MTC), so it's not clear that we absolutely need to handle this case. Maybe we don't worry about it since it can only happen with a misconfigured cluster.

Reply:
@sseago As explained in the previous comment, we are taking defensive approach. We try to calculate the most restrictive values of all. There's one corner case that this code does not handle and that happens when the `LimitRanges` are configured such that it is "impossible" to run Rsync Pods with two containers and by staying in the Min and Max limits of the namespace. In that case, we will fail to create Pods and it will only succeed after the user fixes their `LimitRanges`.